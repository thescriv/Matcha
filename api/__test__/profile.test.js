const superagent = require('superagent')
const webToken = require('jsonwebtoken')
const crypto = require('crypto')

const db = require('../src/lib/db')
const jwt = require('../src/lib/webToken')
const config = require('../../../config')

const { script } = require('../scripts/script')
const { deleteRows } = require('./utils/deleteRows')
const { getRandomPort } = require('./utils/getRandomPort')
const { startApi, stopApi } = require('../api')

let baseUrl

describe(`profile -- `, () => {
  beforeAll(async () => {
    const port = getRandomPort()

    baseUrl = `http://localhost:${port}/api`

    await script('matchaTest')

    await deleteRows()

    await startApi(port)
  })

  afterEach(async () => {
    await deleteRows()

    jest.restoreAllMocks()
  })

  afterAll(async () => {
    await stopApi()
  })

  describe('auth', () => {
    test('do not auth (token cannot be verified)', async () => {
      jest.spyOn(jwt, 'verify').mockImplementation(() => {
        throw new Error('api.generateWT token cannot_be_verified')
      })

      const token = jwt.generate('1')

      let error

      try {
        await superagent
          .get(`${baseUrl}/profile/`)
          .set('Authorization', `Bearer ${token}`)
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })

    test('do not auth (no token)', async () => {
      try {
        await superagent.get(`${baseUrl}/profile/`)
      } catch (err) {
        error = err
      }
      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })

    test('do not auth (user is not logged)', async () => {
      const token = webToken.sign(
        { auth: ['user'], user_id: '1', logged: false },
        config.SECRET_KEY,
        {
          expiresIn: '1h',
        }
      )

      let error

      try {
        await superagent
          .get(`${baseUrl}/profile/`)
          .set('Authorization', `Bearer ${token}`)
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })

    test('do not auth (no user_id in token)', async () => {
      const token = webToken.sign(
        { auth: ['user'], logged: true },
        config.SECRET_KEY,
        {
          expiresIn: '1h',
        }
      )

      let error

      try {
        await superagent
          .get(`${baseUrl}/profile/`)
          .set('Authorization', `Bearer ${token}`)
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })

    test('do not auth (no logged in token)', async () => {
      const token = webToken.sign(
        { auth: ['user'], user_id: '1' },
        config.SECRET_KEY,
        {
          expiresIn: '1h',
        }
      )

      let error

      try {
        await superagent
          .get(`${baseUrl}/profile/`)
          .set('Authorization', `Bearer ${token}`)
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })
  })

  describe('GET /profile/me', () => {
    test('do get me', async () => {
      const insertedUser = await db.query(
        `INSERT INTO user (firstname, lastname, email, password) 
              VALUES ("theo", "test", "test@test.test", "f94b937b40c946b916c63a1b638ee956")`
      )

      const token = jwt.generate(insertedUser.insertId)

      const { body, status } = await superagent
        .get(`${baseUrl}/profile/me`)
        .set('Authorization', `Bearer ${token}`)

      expect({ body, status }).toMatchSnapshot()
    })

    test('do not get me (user does not exist)', async () => {
      const token = jwt.generate('1')

      let error

      try {
        await superagent
          .get(`${baseUrl}/profile/me`)
          .set('Authorization', `Bearer ${token}`)
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })
  })

  describe('GET /profile/logout', () => {
    test('do logout', async () => {
      const token = jwt.generate('1')

      const { body, status } = await superagent
        .get(`${baseUrl}/profile/logout`)
        .set('Authorization', `Bearer ${token}`)

      expect({ body, status }).toMatchSnapshot()
    })

    test('do not logout (jwt.logout throw)', async () => {
      jest.spyOn(jwt, 'logout').mockImplementation(() => {
        throw new Error('api.generateWT token cannot_be_generated')
      })
      const token = jwt.generate('1')

      try {
        await superagent
          .get(`${baseUrl}/profile/logout`)
          .set('Authorization', `Bearer ${token}`)
      } catch (err) {
        error = err
      }
      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()
    })
  })

  describe('POST /profile/update', () => {
    let token

    let userId

    beforeEach(async () => {
      const insertedUser = await db.query(
        `INSERT INTO user (nickname, firstname, lastname, email, password, verified) 
              VALUES ("thescriv", "theo", "test", "test@test.test", "f94b937b40c946b916c63a1b638ee956", true)`
      )

      userId = insertedUser.insertId

      token = jwt.generate(userId)
    })

    test('do update profile', async () => {
      const { body, status } = await superagent
        .post(`${baseUrl}/profile/update`)
        .set('Authorization', `Bearer ${token}`)
        .send({ firstname: 'Jack' })

      expect({ body, status }).toMatchSnapshot()

      const user = await db.query('SELECT * FROM user')
      expect(user).toMatchSnapshot()
    })

    test('do update profile (password modified)', async () => {
      jest
        .spyOn(crypto, 'randomBytes')
        .mockReturnValue('0000000000000000000000000000000000000001')

      const { body, status } = await superagent
        .post(`${baseUrl}/profile/update`)
        .set('Authorization', `Bearer ${token}`)
        .send({ firstname: 'Jack', password: 'Thescriv1' })

      expect({ body, status }).toMatchSnapshot()

      const user = await db.query('SELECT * FROM user')
      expect(user).toMatchSnapshot()
    })

    test('do not update profile (no token send)', async () => {
      let error

      try {
        await superagent
          .post(`${baseUrl}/profile/update`)
          .send({ firstname: 'Jack' })
      } catch (err) {
        error = err
      }
      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const user = await db.query('SELECT * FROM user')
      expect(user).toMatchSnapshot()
    })

    test('do not update profile (user does not exist)', async () => {
      await db.query(`DELETE FROM user WHERE id = ${userId}`)

      let error

      try {
        await superagent
          .post(`${baseUrl}/profile/update`)
          .set('Authorization', `Bearer ${token}`)
          .send({ firstname: 'Jack' })
      } catch (err) {
        error = err
      }
      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const user = await db.query('SELECT * FROM user')
      expect(user).toMatchSnapshot()
    })
  })
})
