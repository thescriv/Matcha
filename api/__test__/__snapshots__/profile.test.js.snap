// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile --  GET /profile/logout do logout 1`] = `
Object {
  "body": Object {},
  "status": 204,
}
`;

exports[`profile --  GET /profile/logout do not logout (jwt.logout throw) 1`] = `
Object {
  "body": Object {
    "error": "api.generateWT token cannot_be_generated",
  },
  "status": 400,
}
`;

exports[`profile --  GET /profile/me do get me 1`] = `
Object {
  "body": Object {
    "profileInfo": Object {
      "bio": null,
      "completed": 0,
      "email": "test@test.test",
      "firstname": "theo",
      "gender_id": null,
      "id": 1,
      "lastname": "test",
      "nickname": null,
      "orientation_id": 1,
      "password": "f94b937b40c946b916c63a1b638ee956",
      "score": 100,
      "verified": 0,
    },
  },
  "status": 200,
}
`;

exports[`profile --  GET /profile/me do not get me (user does not exist) 1`] = `
Object {
  "body": Object {
    "error": "api.profile user does_not_exist",
  },
  "status": 400,
}
`;

exports[`profile --  POST /profile/update do not update profile (no token send) 1`] = `
Object {
  "body": Object {},
  "status": 401,
}
`;

exports[`profile --  POST /profile/update do not update profile (no token send) 2`] = `
Array [
  RowDataPacket {
    "bio": null,
    "completed": 0,
    "email": "test@test.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 1,
  },
]
`;

exports[`profile --  POST /profile/update do not update profile (user does not exist) 1`] = `
Object {
  "body": Object {
    "error": "api.updateProfile user does_not_exist",
  },
  "status": 400,
}
`;

exports[`profile --  POST /profile/update do not update profile (user does not exist) 2`] = `Array []`;

exports[`profile --  POST /profile/update do update profile (password modified) 1`] = `
Object {
  "body": Object {},
  "status": 204,
}
`;

exports[`profile --  POST /profile/update do update profile (password modified) 2`] = `
Array [
  RowDataPacket {
    "bio": null,
    "completed": 1,
    "email": "test@test.test",
    "firstname": "Jack",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "a720c6b5fc70c9bdfc257736b3140108",
    "score": 100,
    "verified": 1,
  },
]
`;

exports[`profile --  POST /profile/update do update profile 1`] = `
Object {
  "body": Object {},
  "status": 204,
}
`;

exports[`profile --  POST /profile/update do update profile 2`] = `
Array [
  RowDataPacket {
    "bio": null,
    "completed": 1,
    "email": "test@test.test",
    "firstname": "Jack",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 1,
  },
]
`;

exports[`profile --  auth do not auth (no logged in token) 1`] = `
Object {
  "body": Object {
    "err": "api.generateWT token user_is_not_logged",
  },
  "status": 400,
}
`;

exports[`profile --  auth do not auth (no token) 1`] = `
Object {
  "body": Object {},
  "status": 401,
}
`;

exports[`profile --  auth do not auth (no user_id in token) 1`] = `
Object {
  "body": Object {
    "err": "auth token unknow_token",
  },
  "status": 401,
}
`;

exports[`profile --  auth do not auth (token cannot be verified) 1`] = `
Object {
  "body": Object {
    "err": "api.generateWT token cannot_be_verified",
  },
  "status": 400,
}
`;

exports[`profile --  auth do not auth (user is not logged) 1`] = `
Object {
  "body": Object {
    "err": "api.generateWT token user_is_not_logged",
  },
  "status": 400,
}
`;

exports[`profile --  auth do not auth (user is not logged) 2`] = `
Object {
  "body": Object {
    "err": "api.generateWT token user_is_not_logged",
  },
  "status": 400,
}
`;
